components/layout_2020/positioned.rs

```diff
@@ -936,25 +951,89 @@ impl<'a> AbsoluteAxisSolver<'a> {
-            _ => None,
+            AlignFlags::SELF_START => match self.axis {
+                AxisDirection::Block => Some(alignment_container.origin),
+                AxisDirection::Inline => {
+                    if wrapper_writing_mode.is_bidi_ltr() == abspos_writing_mode.is_bidi_ltr() {
+                        Some(alignment_container.origin)
+                    } else if wrapper_writing_mode.is_bidi_ltr() &&
+                        !abspos_writing_mode.is_bidi_ltr()
+                    {
+                        Some(
+                            alignment_container.origin + alignment_container.length -
+                                margin_box_axis.length,
+                        )
+                    } else {
+                        Some(
+                            alignment_container.origin + alignment_container.length -
+                                margin_box_axis.length,
+                        )
+                    }
+                },
+            },
+            AlignFlags::SELF_END => match self.axis {
+                AxisDirection::Block => Some(
+                    alignment_container.origin + alignment_container.length -
+                        margin_box_axis.length,
+                ),
+                AxisDirection::Inline => {
+                    if wrapper_writing_mode.is_bidi_ltr() == abspos_writing_mode.is_bidi_ltr() {
+                        Some(
+                            alignment_container.origin + alignment_container.length -
+                                margin_box_axis.length,
+                        )
+                    } else if wrapper_writing_mode.is_bidi_ltr() &&
+                        !abspos_writing_mode.is_bidi_ltr()
+                    {
+                        Some(alignment_container.origin)
+                    } else {
+                        Some(alignment_container.origin)
+                    }
+                },
+            },
+            AlignFlags::LEFT => {
+                if wrapper_writing_mode.is_bidi_ltr() {
+                    Some(alignment_container.origin)
+                } else {
+                    Some(
+                        alignment_container.origin + alignment_container.length -
+                            margin_box_axis.length,
+                    )
+                }
+            },
+            AlignFlags::RIGHT => {
+                if wrapper_writing_mode.is_bidi_ltr() {
+                    Some(
+                        alignment_container.origin + alignment_container.length -
+                            margin_box_axis.length,
+                    )
+                } else {
+                    Some(alignment_container.origin)
+                }
+            },
+            _ => Some(alignment_container.origin),
         }
     }
 
     fn solve_alignment(
         &self,
         margin_box_rect: LogicalRect<Au>,
         content_box_rect: &mut LogicalRect<Au>,
+        wrapper_writing_mode: WritingMode,
```

Comment:
Avoid "wrapper". The right term is "alignment container", which is the "containing block" in this case. "container" may be clear enough, but otherwise it should be one of the 2 full terms, not "wrapper".